File: fixtures/typescript/repomap/handlers.ts
// `validateAndPersistSignup` deliberately mirrors fixtures/javascript's own
// handlers.js: two callers (handleSignupRoute, retryQueueWorker), one of
// which -- the retry path -- has no rate limiting. Also the fixture's
// line-shift case (REQUIREMENTS.md structural requirement 4):
// `validateAndPersistSignup` is immediately followed by `handleSignupRoute`
// and `retryQueueWorker`, so editing inside the first shifts the other two's
// line numbers without changing their content.

import { validateEmail, hashPassword } from './utils';
import { insertUser, findUserByEmail, updateUser, deleteUser } from './db';
import { sendWelcomeEmail, sendPasswordReset, renderTemplate } from './email';
import { logEvent } from './logging';
import type { User, Role } from './models';

interface SignupPayload {
  email: string;
  password: string;
}

interface RouteContext {
  body: SignupPayload;
  params: { id: string };
  query: unknown;
}

const updateCache = new Map<number, { user: User; cachedAt: number }>();

export function validateAndPersistSignup(data: SignupPayload): User {
  validateEmail(data.email);
  hashPassword(data.password);
  const user = insertUser(data as unknown as User);
  sendWelcomeEmail(user);
  logEvent(`signup persisted for ${data.email}`);
  return user;
}

export function handleSignupRoute(ctx: RouteContext): User {
  const user = validateAndPersistSignup(ctx.body);
  logEvent('handled signup route');
  return user;
}

export function retryQueueWorker(job: { data: SignupPayload }): User {
  const user = validateAndPersistSignup(job.data);
  logEvent('retried signup from queue');
  return user;
}

export function handleLoginRoute(ctx: RouteContext): User | null {
  const user = findUserByEmail(ctx.body.email);
  hashPassword(ctx.body.password);
  logEvent('handled login route');
  return user;
}

export function handleUpdateRoute(ctx: RouteContext): unknown {
  const user = updateUser(Number(ctx.params.id), ctx.body);
  updateCache.set(Number(ctx.params.id), { user: user as unknown as User, cachedAt: Date.now() });
  logEvent('handled update route');
  return user;
}

export function handleDeleteRoute(ctx: RouteContext): void {
  deleteUser(Number(ctx.params.id));
  logEvent('handled delete route');
}

export function handlePasswordResetRoute(ctx: RouteContext): void {
  sendPasswordReset(ctx.body as unknown as User);
  logEvent('handled password reset route');
}

export function handleRenderRoute(ctx: RouteContext): string {
  const html = renderTemplate('page', ctx.query);
  logEvent('handled render route');
  return html;
}

export function handleHealthCheck(defaultRole: Role): void {
  logEvent(`handled health check (default role ${defaultRole})`);
}

File: fixtures/python/repomap/handlers.py
# `validate_and_persist_signup` deliberately mirrors the JS/TS fixtures'
# shape (Session 3's original design): two callers, `handle_signup_route` and
# `retry_queue_worker`, immediately following it in this file -- the
# line-shift structural requirement (fixtures/REQUIREMENTS.md #4). Editing
# inside `validate_and_persist_signup` shifts both callers' line numbers
# without changing their content.

from logging import log_event
from utils import validate_email, hash_password
from db import insert_user, find_user_by_email, update_user, delete_user
from email import send_welcome_email, send_password_reset, render_template

update_cache = {}


def validate_and_persist_signup(data):
    validate_email(data["email"])
    hash_password(data["password"])
    user = insert_user(data)
    send_welcome_email(user)
    log_event(f"signup persisted for {data['email']}")
    return user


def handle_signup_route(request):
    user = validate_and_persist_signup(request["body"])
    log_event("handled signup route")
    return user


def retry_queue_worker(job):
    user = validate_and_persist_signup(job["data"])
    log_event("retried signup from queue")
    return user


def handle_login_route(request):
    user = find_user_by_email(request["body"]["email"])
    hash_password(request["body"]["password"])
    log_event("handled login route")
    return user


def handle_update_route(request):
    user = update_user(request["params"]["id"], request["body"])
    update_cache[user["id"]] = user
    log_event("handled update route")
    return user


def handle_delete_route(request):
    delete_user(request["params"]["id"])
    log_event("handled delete route")


def handle_password_reset_route(request):
    send_password_reset(request["body"])
    log_event("handled password reset route")


def handle_render_route(request):
    html = render_template("page", request["query"])
    log_event("handled render route")
    return html


def handle_health_check(request):
    log_event("handled health check")
